horovod/spark/common/util.py

```diff
@@ -0,0 +1,518 @@
+            col_size = metadata[col]['shape']
+            input_size = abs(np.prod(input_shape))
+            if col_size != input_size:
+                raise ValueError(
+                    'Feature column \'{col}\' with size {feature} must equal that of the '
+                    'model input at index {idx} with size {input}'
+                    .format(col=col, feature=col_size, idx=idx, input=input_size))
+
+    label_count = len(label_columns)
+    if output_shapes is not None and all(metadata[col]['shape'] for col in label_columns):
+        if label_count != len(output_shapes):
+            raise ValueError('Label column count {labels} must equal '
+                             'model outputs count {outputs}'
+                             .format(labels=label_count, outputs=len(output_shapes)))
+
+        for idx, col, output_shape in zip(range(label_count), label_columns, output_shapes):
+            col_size = metadata[col]['shape']
+            output_size = abs(np.prod(output_shape))
+            if col_size != output_size:
+                raise ValueError('Label column \'{col}\' with size {label} must equal that of the '
+                                 'model output at index {idx} with size {output}'
+                                 .format(col=col, label=col_size, idx=idx, output=output_size))
+
+
+def _get_col_info(df):
+    """
+    Infer the type and shape of all the columns.
+    """
+
+    def get_meta(row):
+        row_dict = row.asDict()
+        row_schema = []
+        for col_name, data_col in row_dict.items():
+            dtype = type(data_col)
+            if dtype == DenseVector:
+                # shape and size of dense vector are the same
+                shape = data_col.array.shape[0]
+                size = shape
+            elif dtype == SparseVector:
+                # shape is the total size of vector
+                shape = data_col.size
+                # size is the number of nonzero elements in the sparse vector
+                size = data_col.indices.shape[0]
+            else:
+                shape = 1
+                size = 1
+            row_schema.append((col_name, (set([dtype]), set([shape]), set([size]))))
+        return row_schema
+
+    def merge(x, y):
+        dtypes = x[0]
+        dtypes.update(y[0])
+        shapes = x[1]
+        shapes.update(y[1])
+        sizes = x[2]
+        sizes.update(y[2])
+        return dtypes, shapes, sizes
+
+    raw_col_info_list = df.rdd.flatMap(get_meta).reduceByKey(merge).collect()
+
+    all_col_types = {}
+    col_shapes = {}
+    col_max_sizes = {}
+
+    for col_info in raw_col_info_list:
+        col_name = col_info[0]
+
+        all_col_types[col_name] = col_info[1][0]
+        col_shapes[col_name] = col_info[1][1]
+        col_max_sizes[col_name] = col_info[1][2]
+
+    # all the rows of each columns must have the same shape
+    for col in df.schema.names:
+        shape_set = col_shapes[col]
```

Comment:
if there is a column that has only `None`s, then `col_shapes` might not contain it

Reply:
If the column is all `None`, then it will still have size and shape of 1.  Added tests to verify this.